### ISSUE TITLE
Cannot use Globally Defined Python Environment While Inside Qtile

### ISSUE THREAD

**Issue by jokatzke** [NixOS/nixpkgs#186243]
### Describe the bug
This is almost the same problem as described in #167328, but it only occurs while inside Qtile.

### Steps To Reproduce
1. Declaratively define a global Python environment in ```configuration.nix```:
  ```nix
  environment.systemPackages = with pkgs; [
   (let
      my-python-packages = python-packages: with python-packages; [
        matplotlib
        numpy
        pandas
        scikit-learn
      ];
      python-with-my-packages = python3.withPackages my-python-packages;
    in
      python-with-my-packages)
];
```
2. Test the Python environment
```
$ python -c "import pandas"
```
3. Enable Qtile:
```nix
services.xserver.windowManager = {
  qtile = {
    enable = true;
  };
};
```
4. Log into Qtile
5. Run the test again
```
$ python -c "import pandas"
Traceback (most recent call last):
 File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'pandas'
```

### Expected behavior
I expected to be able to use the globally defined Python environment, regardless of which window manager I am currently using.

### Notify maintainers
@jonringer @kamilchm 

### Metadata
```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.59, NixOS, 22.05 (Quokka), 22.05.2361.45c9736ed69`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.8.1`
 - channels(root): `"home-manager-22.05.tar.gz, nixos-22.05"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

**Comment by nieomylnieja** [comment]
@arjan-s do you think it's possible now to go around this issue with the change you've made here? https://github.com/NixOS/nixpkgs/pull/220450

**Comment by arjan-s** [comment]
I have not tested this yet because I don't globally install Python packages and additionally I start qtile differently. But I can see whether I can make some time for it. If you could confirm whether this is still a problem with current nixos-unstable, that would be great.

**Comment by DCsunset** [comment]
I can confirm that it's still a problem. I found that the python's path is different in Qtile and tty, thus resulting in different environments. I think you can verify it by running `realpath -e $(which python)`.

**Comment by nieomylnieja** [comment]
@arjan-s could you share your way of starting Qtile? If doing it your way would solve the issue it would be great :)

**Comment by arjan-s** [comment]
I'm afraid my setup has the same behaviour as yours, so that won't help...
Also I did a quick test with the solution from https://github.com/NixOS/nixpkgs/pull/146290, but that didn't work. Hopefully I can make some time soon to really try to resolve this.

**Comment by nieomylnieja** [comment]
Thank you so much for the involvement in this! At least now I know there might not be any way to go around it for now :)

**Comment by arjan-s** [comment]
Yeah it seems to be a widespread problem, if you look especially at the comments in the other issues.

Whether it's actually a problem depends, as it's just normal behaviour for child processes to inherit `$PATH` from their parent. But in this case, where Qtile is wrapped inside a separate Python environment, that means that Python occurs earlier in `$PATH` than the system Python.

I think you have a few options to explore:

- Make your terminal emulator clean up your PATH like in https://github.com/NixOS/nixpkgs/issues/171972#issuecomment-1517324462 , although this probably doesn't work if you use something like VScode
- Install the nonwrapped Qtile python package globally: add `qtile` and its dependencies to your `my-python-packages`, then remove `services.xserver.windowManager.qtile` and make your own session start script using `services.xserver.windowManager.session` ([example here](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/x11/window-managers/qtile.nix#L54C5-L54C43))
- Add the system Python's `site-packages` directory to your `$PYTHONPATH` from something like `.bashrc`
- Use something like Poetry so you can easily work with virtual environments instead of your system Python
- Use `su - <youruser>` to get a shell with clean `$PATH` before using Python

**Comment by Rouzihiro** [comment]
Can someone share his qtile configuration? Would love to get it running under NixOS

**Comment by arjan-s** [comment]
@Rouzihiro the easiest way would be to use the qtile display manager session, for instance like this (copied from my own config):
```
security = {
  pam.services = {
    swaylock = { };
    login.enableGnomeKeyring = true;
  };
  polkit.enable = true;
};
hardware.graphics.enable = true;
xdg.portal = {
  enable = true;
  wlr.enable = lib.mkForce true;
  config.common.default = "wlr";
  extraPortals = [ pkgs.xdg-desktop-portal-gtk ];
};  
programs = {
  xwayland.enable = true;
  dconf.enable = true;
  gdk-pixbuf.modulePackages = [ pkgs.librsvg ];
};
services = {
  displayManager.sddm.enable = true;
  xserver.windowManager.qtile = {
    enable = true;
    extraPackages = pp: with pp; [ qtile-extras ];
  };
};
```

**Comment by misuzu** [comment]
This did the trick for me
```nix
{
  nixpkgs.overlays = [
    (self: super: {
      python3 = super.python3.override {
        packageOverrides = python-self: python-super: {
          qtile = python-super.qtile.overrideAttrs (oldAttrs: {
            dontWrapPythonPrograms = true;
            postFixup = ''
              wrapPythonPrograms
              # get rid of PATH magic after wrapper script
              mv $out/bin/.qtile-wrapped $out/bin/qtile
            '';
          });
        };
      };
    })
  ];
}
```

It broke type checking but I can live with that (edit: looks like it didn't work from the beginning)
```
% qtile check
Checking Qtile config at: /home/misuzu/.config/qtile/config.py
Checking if config is valid python...
mypy was not found in PATH. Please install it, add to PATH and try again.
stubtest was not found in PATH. Please install it, add to PATH and try again.
Missing dependencies. Skipping type checking.
```
